src/IO/WriteBuffer.h

```diff
@@ -100,11 +101,26 @@ class WriteBuffer : public BufferBase
         next();
     }
 
-    virtual void finalize()
+    /// Write the last data.
+    void finalize()
+    {
+        if (finalized)
+            return;
+
+        /// finalize() is often called from destructors.
+        MemoryTracker::LockExceptionInThread lock(VariableContext::Global);
+        finalized = true;
+        finalizeImpl();
```

Comment:
Can we set `pos = working_buffer.begin()` in case of exception just for every buffer?

Reply:
Also maybe check that buffer is not finalized before every write operation and throw?